Add ability to enroll with a specific ID

blakerouse opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What is the problem this PR solves?

This solves an issue where an Elastic Agent is being replaced with a new Elastic Agent instance for the same host, pod, or workload. This allows the enrolling Elastic Agent to tell the ID that it wants to use, that ID can be currently in-use and this enrollment will take over the record of that Elastic Agent.

## How does this PR solve the problem?

It solves the issue by taking a new `id` field in the enroll HTTP request. That `id` is then used as the Elastic Agent ID and determines if this is a new Elastic Agent or if it should take over an existing Elastic Agent record.

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer if anything special is needed for manual testing: commands, dependencies, steps, etc.
-->

At the moment the integration tests are the best way to test this, as the ability to use this field has not been exposed yet on the Elastic Agent.

## Design Checklist

<!-- Mandatory
This checklist is a reminder about high level design problems that should be considered for any change made to fleet server.
-->

- [x] I have ensured my design is stateless and will work when multiple fleet-server instances are behind a load balancer.
- [x] I have or intend to scale test my changes, ensuring it will work reliably with 100K+ agents connected.
- [x] I have included fail safe mechanisms to limit the load on fleet-server: rate limiting, circuit breakers, caching, load shedding, etc. (already covered by enroll handle)

## Checklist

<!-- Mandatory
This checklist is to help creators of PRs to find parts which might not be directly related to code change but still need to be addressed. Anything that does not apply to the PR should be removed from the checklist.
-->

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- ~~[ ] I have made corresponding change to the default configuration files~~ (no config changes)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/fleet-server#changelog)


## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes #4226

jlind23 commented:
After chatting with @blakerouse over slack we should fail the enrollment of this "new" agent if the policy has temper protection enabled.

kaanyalti commented:
Changes so far look good to me, waiting for updates relevant to @jlind23's comment above to approve

michel-laterman commented on model/openapi.yml:
```diff
@@ -139,6 +139,13 @@ components:
+          description: |
+            The ID of the agent.
+            This is the ID that will be used to reference this agent, if no ID is passed one will be generated.
+            If another agent is enrolled with the same ID the other agent will no longer be able to communicate,
+            this new agent is considered a replacement of the other agent.
```
This is a pretty big change from our discussion.
Please also add a sentence saying the (replaced) agent will still be able to send data into ES

blakerouse replied:
It is a slight change, because it is not possible to get an API key token again after the initial create. That made me have no choice to change the behavior.

blakerouse replied:
I did add more as requested to the description of this field to inform that it will still allow data to flow.

blakerouse commented:
@jlind23 @michel-laterman @kaanyalti I have updated this PR based on the discuss I had with @jlind23 about security with this feature. This PR now includes an additional `replace_token` during the enrollment API. I have updated the PR description to describe this as well as the API specification describes it.

pkoutsovasilis commented on internal/pkg/api/handleEnroll.go:
```diff
@@ -252,6 +260,105 @@ func (et *EnrollerT) _enroll(
+				zlog.Error().Err(err).
+					Str("AgentId", agent.Id).
+					Msg("Error when trying to delete old agent with same id")
+				return nil, err
```
missing `vSpan.End()` right?

michalpristas replied:
we could extract to separate function that finds agent id with defer vspan.End to decrease probability of this in a future

pkoutsovasilis commented:
Except a small ending of a trace-span the code changes look good to me. I understand the potential pitfalls with this feature and definitely see how the `replace_token` helps in minimising some of them but still this feature to me serves only special-case scenarios and is not streamlined usage 🙂

michalpristas commented on internal/pkg/api/handleEnroll.go:
```diff
@@ -197,10 +201,22 @@ func (et *EnrollerT) _enroll(
+
+	if req.ReplaceToken != nil && *req.ReplaceToken != "" {
+		replaceTokenBytes, err := bcrypt.GenerateFromPassword([]byte(*req.ReplaceToken), bcrypt.DefaultCost)
+		if err != nil {
```
should we log error, looking at implementation of GenerateFromPassword it can be miriad of things

blakerouse replied:
Added a log message.

michalpristas commented on internal/pkg/api/handleEnroll.go:
```diff
@@ -252,6 +260,105 @@ func (et *EnrollerT) _enroll(
+				return nil, ErrAgentNotReplaceable
+			}
+			err = bcrypt.CompareHashAndPassword([]byte(agent.ReplaceToken), []byte(*req.ReplaceToken))
+			if err != nil {
```
could we log error in case `err != bcrypt.ErrMismatchedHashAndPassword`

blakerouse replied:
Added log.

michalpristas commented on internal/pkg/api/handleEnroll.go:
```diff
@@ -252,6 +260,105 @@ func (et *EnrollerT) _enroll(
+					Str("PolicyId", policyID).
+					Str("CurrentPolicyId", agent.PolicyID).
+					Msg("Existing agent with same ID already enrolled into another policy")
+				return nil, ErrAgentInAnotherPolicy
```
very minor: i wonder if this is not telling too much, attacker could try agentID and token. this would tell token is correct, policy is not. Maybe returning ErrAgentNotReplaceable with log unchanged would make more sense?

blakerouse commented:
@pkoutsovasilis @michalpristas I updated the PR with the request fixes. Thanks for the reviews.

pkoutsovasilis commented on internal/pkg/api/handleEnroll.go:
```diff
@@ -271,47 +355,84 @@ func (et *EnrollerT) _enroll(
+			dl.FieldAuditUnenrolledReason: nil,
+			dl.FieldUnenrolledAt:          nil,
+			dl.FieldUnenrolledReason:      nil,
+			dl.FieldUpdatedAt:             time.Now().UTC().Format(time.RFC3339),
```
nit; here we use `time.Now()` while below we use the top declared variable `now`, is there a reason these two are using the same time?

blakerouse replied:
Nice catch, missed that it was above. I updated it to use the `now` variable from above.

pkoutsovasilis reviewed (approved):
LGTM

michalpristas commented on internal/pkg/api/handleEnroll.go:
```diff
@@ -252,6 +261,81 @@ func (et *EnrollerT) _enroll(
+			return nil, err
+		}
+
+		if agent.Id != "" {
```
can it return `agent.id == ""`? if not this if statement is not needed.
if so. we dont have this case handled as this should not be the same as empty ID when `req.ID` is not used.
we whould probably not continue with empty id, probably we should fail. generating a new one breaks the purpose of providing it via req.id

blakerouse replied:
It absolutely will return an `agent.id == ""`. This happens when we check to see if an existing agent already exists with that ID. The ErrNotFound will not be returned from `_checkAgent`, it will return `nil` error and this will be `agent.id == "".

michalpristas reviewed (approved):
Thanks for resolving comment. one question about behavior other than that I'm ok with the change. 
i let you decide how you want to address the point i raised in this iteration

cmacknz commented on internal/pkg/api/handleEnroll.go:
```diff
@@ -197,10 +200,24 @@ func (et *EnrollerT) _enroll(
+	now := time.Now()
+
+	if req.ReplaceToken != nil && *req.ReplaceToken != "" {
+		replaceTokenBytes, err := bcrypt.GenerateFromPassword([]byte(*req.ReplaceToken), bcrypt.DefaultCost)
```
I don't think the blowfish hash in the underlying implementation here meets FIPS 140-2. At least for the MS Go fork it wouldn't hit the FIPS certified OpenSSL so would be non-compliant. The list of what gets forwarded to it is at https://github.com/microsoft/go/blob/microsoft/main/eng/doc/fips/UserGuide.md

Whatever crypto primitives we use need to be FIPS compliant.

CC @michel-laterman

pkoutsovasilis replied:
That's actually more than a valid point @cmacknz 👍 AFAIK the only FIPS compliant password hashing is PBKDF2. I had a look and all call in it are forwarded by MS Go fork. We already use PBKDF2 password hashing in [elastic-agent](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/crypto/io.go#L376) and I believe we should do the same here as well

cmacknz reviewed (changes requested):
Requesting changes until we get a FIPS compliant implementation. We're blocked on security review anyway.

blakerouse commented:
@cmacknz @pkoutsovasilis Thanks for catching this now for FIPS. I have updated the implementation to use PBKDF2 and SHA512 like the Elastic Agent. Let me know what you think.

pkoutsovasilis commented on internal/pkg/api/handleEnroll.go:
```diff
@@ -538,3 +711,56 @@ func validateRequest(ctx context.Context, data io.Reader) (*EnrollRequest, error
+		return false, ErrAgentCorrupted
+	}
+	key := pbkdf2.Key([]byte(token), salt, iterations, cfg.KeyLength, sha512.New)
+	return bytes.Equal(key, encoded), nil
```
```suggestion
	return hmac.Equal(key, encoded), nil
```

blakerouse replied:
Why `hmac.Equal`? Whats the reason? Would like to add that in the code comment so its clear. Its not clear to my why that is used here.

blakerouse replied:
Got it, thanks for the details. I have updated it to use `hmac.Equal` and added a comment.

pkoutsovasilis commented on internal/pkg/config/pbkdf2.go:
```diff
@@ -0,0 +1,42 @@
+	KeyLength  int `config:"key_length"`
+	SaltLength int `config:"salt_length"`
+
+	// BlockSize must be a factor of aes.BlockSize
+	BlockSize int `config:"block_size"`
```
```suggestion
```
is this actually used somewhere?

blakerouse replied:
I copied the configuration from Elastic Agent. I will remove that as that is not used here.

blakerouse replied:
Removed

pkoutsovasilis commented on internal/pkg/config/pbkdf2.go:
```diff
@@ -0,0 +1,42 @@
+// InitDefaults is the default options to use with PDKDF2, changing might decrease
+// the efficacy of the encryption.
+func (p *PBKDF2) InitDefaults() {
+	p.Iterations = 10000
```
fact: In 2023, OWASP recommended to use [210,000](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pbkdf2) for PBKDF2-HMAC-SHA512. The more iterations the more time it will take to compute the hash, but not quite sure if we should follow this advise or not 🙂

blakerouse replied:
@pkoutsovasilis I was just using the defaults that the Elastic Agent uses so they can be inline. Should we also update the Elastic Agent.

https://github.com/elastic/elastic-agent/blob/main/internal/pkg/crypto/io.go#L58

pkoutsovasilis replied:
yy I ve seen that, however, I am not entirely sure how easy of a change is for the elastic-agent without breaking anything. Need to look into that 🙂

blakerouse replied:
Updated here at least.

cmacknz commented on model/openapi.yml:
```diff
@@ -139,6 +139,14 @@ components:
+            The ID of the agent.
+            This is the ID that will be used to reference this agent, if no ID is passed one will be generated.
+            If another agent is enrolled with the same ID the other agent will no longer be able to communicate,
+            this new agent is considered a replacement of the other agent. The other agent will be able to continue
```
> The other agent will be able to continue sending data to ES.

How do the ES API keys eventually get revoked? This normally happens through unenrolling an agent via the UI, but with replacement of the agent, you lose that.

Should replacing an agent revoke it's ES API keys to treat it like a force unenroll but triggered through fleet server?

cmacknz replied:
I think I maybe misunderstood what happens the first time, the ES API keys are tied to the agent ID so they are kept alive because they are reused.

So this is documenting that this feature only works as expected if the agent you are replacing stops running. Otherwise you will have two agents ingesting data but one of them is unmanaged and has had its Fleet API key revoked. Correct?

blakerouse replied:
That is correct. It will be able to keep ingesting data, which we want for a short period, but will not be able to communicate with Fleet. Then its up to the deployer to turn it off when the other one is working.

This works for Kubernetes, as it will allow a new pod to spawn while the previous version is still running. Once the new version is up and running and healthy then the other pod will be stopped.

cmacknz commented on internal/pkg/api/handleEnroll.go:
```diff
@@ -252,6 +253,85 @@ func (et *EnrollerT) _enroll(
+				Str("AgentId", agent.Id).
+				Str("APIKeyID", agent.AccessAPIKeyID).
+				Msg("Invalidate old api key with same id")
+			err = invalidateAPIKey(ctx, zlog, et.bulker, agent.AccessAPIKeyID)
```
Does any of the state in the .fleet-agents document need to be reset? For example the new agent is not going to have the same persisted action state or .fleet-action sequence number, and will not yet have a copy of the policy. 

If we do any state resets there, it has to happen after the API key confirmed invalidated and not just when it is requested to avoid race conditons.

blakerouse replied:
The `.fleet-action` sequence number I do not reset, as I don't think we want this Elastic Agent to replay all actions of the previous Elastic Agent. If this is used outside of the container use-case, you would not expect it to roll through performing an upgrade action because it replaced a previous Elastic Agent.

I took this into account and didn't reset the sequence number for that exact reason. The policy and revision is not tied to the sequence number in anyway. It is actually stored directly on the `.fleet-agents` document, which is reset by this code here - https://github.com/elastic/fleet-server/pull/4290/files#diff-b3d2d3a49ef214fba08e3aa10772162cbb65c378e24ef6c4a0c5df65b051f71aR378

cmacknz commented on internal/pkg/api/handleEnroll.go:
```diff
@@ -252,6 +253,85 @@ func (et *EnrollerT) _enroll(
+
+			// confirm that its on the same policy
+			// it is not supported to have it the same ID enroll into different policies
+			if agent.PolicyID != policyID {
```
There is an unlikely race condition where a user assigns the agent to a new policy as this replacement is happening. The Fleet server state isn't guaranteed to be current and we don't have any kind of lock or transaction in the index to help us avoid this type of problem.

cmacknz replied:
A user could unenroll the agent as this function is executing to enroll it again with a new API key is another weird situation that could arise.

blakerouse replied:
I don't think we really document specific flows in Fleet Server. Those flows I believe are normally document publicly as Elastic Agent documentation. Let me know where you think I should put this information and I can do that.

cmacknz reviewed (commented):
The implementation LGTM, seems FIPS compliant AFAIK. You could use the Go 1.24 release candidate to double check this, it has a mode that causes non-FIPS crypto to return errors.

I think there are some interesting but unlikely concurrency problems that could be introduced here around:
1. The agent you are replacing being in the process of checking in as you are replacing it. Possibly this gives it a chance to update .fleet-agents information on you. I'm not sure this matters.
2. User actions on the UI side of the .fleet-agents index being in conflict with the re-enrollment here. The user unenrolling the agent you are replacing concurrently is probably the most interesting case.
3. The new, replaced agent having it's action state reset to zero.  Maybe this doesn't matter, but we should test it. Will the agent resetting it's action sequence number to 0 have any unexpected effects?

levinebw commented:
noting, infosec security review discussion here, https://github.com/elastic/fleet-server/issues/4226

blakerouse commented:
Most of my responses to your inline comments are related here, but I will re-iterate.

> 
> 1. The agent you are replacing being in the process of checking in as you are replacing it. Possibly this gives it a chance to 
update .fleet-agents information on you. I'm not sure this matters.

Luckily the way Fleet Server works it only updates fields on a document and not a whole document. Because of this I am not worried as much of all the fields being reset to the wrong value for check-in. Looking at the code the following fields are updated on check-in:

https://github.com/elastic/fleet-server/blob/main/internal/pkg/checkin/bulk.go#L227

I don't see any conflicts there that would cause an issue for the newly replaced Elastic Agent to be able to successfully check-in and update that information. The previous Elastic Agent that just checked-in will not be able from that point forward.

> 2. User actions on the UI side of the .fleet-agents index being in conflict with the re-enrollment here. The user unenrolling the agent you are replacing concurrently is probably the most interesting case.

Not an issue in agentless, as it cannot be unenrolled. I do think outside of agentless there is a small window where this could get in a bad state. I think it is a very small window, but I don't know if the effort involved to prevent that window completely is worth it currently.

> 3. The new, replaced agent having it's action state reset to zero.  Maybe this doesn't matter, but we should test it. Will the agent resetting it's action sequence number to 0 have any unexpected effects?

This is not reseting the action state, see my inline comment. It is just reseting the policy revision to ensure that on its first check-in that it gets it latest policy. I think if we solve this issue here, we don't even need to reset this here - https://github.com/elastic/elastic-agent/issues/6446
